### PR TITLE
Adding unit tests for resolution of issue #470

### PR DESCRIPTION
**Pull request by awickens**
Unit tests added to show issue #470.

### CODE REVIEW

**Comment by awickens** [comment]
Ready for review.

File: src/units.cpp [hunk 1]
```diff
@@ -410,11 +410,13 @@ size_t Units::unitCount() const
 double Units::scalingFactor(const UnitsPtr &units1, const UnitsPtr &units2)
 {
     if ((units1 != nullptr) && (units2 != nullptr)) {
-        double multiplier = 0.0;
+        if ((units1->unitCount() != 0) && (units2->unitCount() != 0)) {
```
Comment (agarny) [inline src/units.cpp:413]: No need as such for nested `if` statements. IOW, I would rework the code to read something like:

```
if ((units1 != nullptr) && (units2 != nullptr)
    && (units1->unitCount() != 0) && (units2->unitCount() != 0)) {
...
}
```

File: tests/units/units.cpp [hunk 2]
```diff
@@ -730,6 +730,31 @@ TEST(Units, compareScalingFactorWithNullptrAsBothParameters)
     EXPECT_EQ(0.0, libcellml::Units::scalingFactor(nullptr, nullptr));
 }
 
+TEST(Units, scalingFactorWithOneUnitsPtrEmpty)
```
Comment (agarny) [inline tests/units/units.cpp]: I woudl rename the test to `scalingFactorWithOneEmptyUnit`.

File: tests/units/units.cpp [hunk 3]
```diff
@@ -730,6 +730,31 @@ TEST(Units, compareScalingFactorWithNullptrAsBothParameters)
...
+    u2->addUnit(libcellml::Units::StandardUnit::LITRE, 0, 1.0, 1.0);
+
+    EXPECT_EQ(0.0, libcellml::Units::scalingFactor(u1, u2));
+    EXPECT_EQ(0.0, libcellml::Units::scalingFactor(u2, u1));
+}
+
+TEST(Units, scalingFactorWithBothUnitsPtrEmpty)
+{
+    libcellml::UnitsPtr u1 = libcellml::Units::create();
+    u1->setName("u1");
```
Comment (agarny) [inline tests/units/units.cpp:778]: I woudl rename the test to `scalingFactorWithTwoEmptyUnits`.

**Review by agarny — CHANGES_REQUESTED** [review]
Looks good, pending some suggestions for some minor "improvements".

**Review by agarny — DISMISSED** [review]

File: src/units.cpp [hunk 4]
```diff
@@ -409,7 +409,8 @@ size_t Units::unitCount() const
 
 double Units::scalingFactor(const UnitsPtr &units1, const UnitsPtr &units2)
 {
-    if ((units1 != nullptr) && (units2 != nullptr)) {
+    if ((units1 != nullptr) && (units2 != nullptr)
+        && (units1->unitCount() != 0) && (units2->unitCount() != 0)) {
```
Comment (hsorby) [inline src/units.cpp]: I have added two tests to show why this condition doesn't work for me.

**Review by hsorby — CHANGES_REQUESTED** [review]

File: src/units.cpp (hunk 4 above)
Comment (awickens) [reply src/units.cpp]: Yep, makes sense. Just added changes which pass both of those tests.

**Comment by awickens** [comment]
Ready for re-review.

**Review by agarny — APPROVED** [review]

**Review by hsorby — APPROVED** [review]

**Comment by agarny** [comment]
Just noticed that this PR  is actually out-of-date with `develop`...

**Pull request merged**